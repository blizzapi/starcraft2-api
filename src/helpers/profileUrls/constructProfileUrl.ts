import BlizzAPI from 'blizzapi';
import checkIfProfileIdLooksValid from '../profileIds/checkIfProfileIdLooksValid';
import validateProfileUrlLocale from '../profileUrlLocales/validateProfileUrlLocale';
import {
  PlayerObject,
} from '../../types';

export default (
  {
    regionId,
    realmId,
    profileId,
  }: PlayerObject,
  localeName?: string,
) => {
  try {
    const validRegionId = BlizzAPI.validateRegionId(regionId);
    const validRealmId = BlizzAPI.checkIfSc2RealmLooksValid(realmId);
    const validProfileId = checkIfProfileIdLooksValid(profileId);
    const locale = localeName && validateProfileUrlLocale(localeName)
      ? localeName
      : 'en-us';

    if (validRegionId && validRealmId && validProfileId) {
      return `https://starcraft2.com/${locale}/profile/${regionId}/${realmId}/${profileId}`;
    }
    return '';
  } catch (error) {
    return '';
  }
};
