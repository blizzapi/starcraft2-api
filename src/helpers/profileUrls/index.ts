export { checkIfProfileUrlLooksValid } from "./checkIfProfileUrlLooksValid";
export { constructProfileUrl } from "./constructProfileUrl";
export { profileUrlRegex } from "./profileUrlRegex";
export { unpackProfileUrl } from "./unpackProfileUrl";
export { validateProfileUrl } from "./validateProfileUrl";
