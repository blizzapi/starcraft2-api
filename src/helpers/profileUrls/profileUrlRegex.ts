/**
 * Exports a regular expression for StarCraft2.com profile URLs
 */
export const profileUrlRegex = /starcraft2\.com\/([A-z]{2}-[A-z]{2})\/profile\/(1|2|3|5{1})\/(1|2{1})\/([0-9]\d*)/i;
