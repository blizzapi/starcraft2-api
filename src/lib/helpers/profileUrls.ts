import BlizzAPI, {
  RegionIdAsNumberOrString,
  Sc2RealmAsNumberOrString,
} from 'blizzapi';

export interface PlayerProfile {
  regionId: RegionIdAsNumberOrString,
  realmId: Sc2RealmAsNumberOrString,
  profileId: string | number,
}

const profileUrlRegex: RegExp =
  /starcraft2\.com\/([A-z]{2}-[A-z]{2})\/profile\/(1|2|3|5{1})\/([0-9]{1})\/([0-9]*)/gi;

export const validateProfileUrl = (url: string) =>
  profileUrlRegex.test(url);

export const validateProfileId = (profileId: number | string) => 
  profileId.toString() === parseInt(profileId as string, 10).toString();

export const unpackProfileUrl = (url: string, includeLocale?: boolean) => {
  const urlIsValid = validateProfileUrl(url);

  if (!urlIsValid) return {};

  const profileUrl = url.match(profileUrlRegex)![0];
  const profileDataArray = profileUrlRegex.exec(profileUrl)!;

  // tslint:disable-next-line: no-object-mutation
  profileUrlRegex.lastIndex = 0;

  const playerObject = {
    regionId: profileDataArray[2],
    realmId: profileDataArray[3],
    profileId: profileDataArray[4],
  };

  return includeLocale
    ? {
      locale: profileDataArray[1],
      ...playerObject,
    }
    : playerObject;
}

export const constructProfileUrl = ({ regionId, realmId, profileId }: PlayerProfile, localeName?: string) => {
  try {
    const validRegionId = BlizzAPI.validateRegionId(regionId);
    const validProfileId = validateProfileId(profileId);
    const validRealmId = BlizzAPI.checkIfSc2RealmLooksValid(realmId);
  
    if (validRegionId && validRealmId && validProfileId) {
      const locale = localeName || 'en-us';
      return `https://starcraft2.com/${locale}/profile/${regionId}/${realmId}/${profileId}`;
    }
    return '';
  } catch (error) {
    return '';
  }
}