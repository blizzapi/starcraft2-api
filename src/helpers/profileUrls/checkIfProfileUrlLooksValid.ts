import profileUrlRegex from './profileUrlRegex';

/**
 * Verifies whether profile URL locale matches regex pattern
 *
 * @param profileUrl Profile URL locale (e.g. 'en-us')
 * @return true if profile URL matches the pattern, false if not
 */
export default (profileUrl: string) =>
  profileUrlRegex.test(profileUrl);
