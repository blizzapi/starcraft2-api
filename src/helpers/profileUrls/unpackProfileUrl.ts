import checkIfProfileUrlLooksValid from './checkIfProfileUrlLooksValid';
import profileUrlRegex from './profileUrlRegex';
import { PlayerObject } from '../../types.d';

export default (url: string, includeLocale?: boolean): PlayerObject | {} => {
  const urlIsValid = checkIfProfileUrlLooksValid(url);

  if (!urlIsValid) return {};

  const profileUrl = url.match(profileUrlRegex)![0];
  const profileDataArray = profileUrlRegex.exec(profileUrl)!;

  return {
    ...includeLocale && { locale: profileDataArray[1] },
    regionId: profileDataArray[2],
    realmId: profileDataArray[3],
    profileId: profileDataArray[4],
  };
};
