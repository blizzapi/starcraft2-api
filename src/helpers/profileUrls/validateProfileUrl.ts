import { BlizzAPI } from "blizzapi";
import { checkIfProfileIdLooksValid } from "../profileIds/checkIfProfileIdLooksValid";
import { validateProfileUrlLocale } from "../profileUrlLocales/validateProfileUrlLocale";
import { unpackProfileUrl } from "./unpackProfileUrl";
import { PlayerObject } from "../../types";

export const validateProfileUrl = (url: string, includeLocale?: boolean) => {
  const profileObject = unpackProfileUrl(url, includeLocale) as PlayerObject;

  if (Object.keys(profileObject).length === 0) {
    throw new RangeError(
      `${url} is not a valid argument for validateProfileUrl()`
    );
  }

  const { regionId, realmId, profileId } = profileObject;

  const localeName = includeLocale
    ? (profileObject as PlayerObject & { locale: string }).locale
    : undefined;

  const validRegionId = BlizzAPI.validateRegionId(regionId);
  const validRealmId = BlizzAPI.checkIfSc2RealmLooksValid(realmId);
  const validProfileId = checkIfProfileIdLooksValid(profileId);
  const validLocale = localeName && validateProfileUrlLocale(localeName);

  return validRegionId && validRealmId && validProfileId && includeLocale
    ? validLocale
    : true;
};
