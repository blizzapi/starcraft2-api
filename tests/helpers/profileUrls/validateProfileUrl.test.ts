import { validateProfileUrl } from "../../../src/helpers/profileUrls/validateProfileUrl";
import validProfileUrls from "../../__testData__/profileUrls.json";
import invalidProfileUrls from "../../__testData__/invalidProfileUrls.json";

describe("validateProfileUrl", () => {
  validProfileUrls.forEach((validProfileUrl) => {
    it(`should match snapshot for valid URL ${validProfileUrl}`, () => {
      expect(validateProfileUrl(validProfileUrl)).toMatchSnapshot();
    });
  });

  validProfileUrls.forEach((validProfileUrl) => {
    it(`should match snapshot for valid URL ${validProfileUrl} with locale`, () => {
      expect(validateProfileUrl(validProfileUrl, true)).toMatchSnapshot();
    });
  });

  invalidProfileUrls.forEach((invalidProfileUrl) => {
    it(`should throw RangeError for invalid URL ${invalidProfileUrl}`, () => {
      expect(() => validateProfileUrl(invalidProfileUrl)).toThrow(RangeError);
    });
  });
});
