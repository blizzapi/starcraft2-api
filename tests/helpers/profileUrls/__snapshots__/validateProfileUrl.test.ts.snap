// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateProfileUrl should match snapshot for valid URL //starcraft2.com/pl-pl/profile/1/1/1084304 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL //starcraft2.com/pl-pl/profile/1/1/1084304 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/1084304 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/1084304 2`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/1084304 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/1084304 with locale 2`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/123456789 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/1/1/123456789 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/2/1/234567890 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/2/1/234567890 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/3/1/345678901 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/3/1/345678901 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/5/1/456789012 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL https://starcraft2.com/en-us/profile/5/1/456789012 with locale 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL starcraft2.com/ru-ru/profile/1/1/1084304 1`] = `true`;

exports[`validateProfileUrl should match snapshot for valid URL starcraft2.com/ru-ru/profile/1/1/1084304 with locale 1`] = `true`;
